docs/question/server.md

```diff
@@ -139,11 +254,86 @@ def file(data):
     # As an example, we will generate the "fig.png" figure.
 
-    if data["filename"] == "fig.png":                # check for the appropriate filename
-        plt.plot([0, data["params"]["a"]], [0, 1])   # plot a line with slope "a"
-        buf = io.BytesIO()                           # make a bytes object (a buffer)
-        plt.savefig(buf, format="png")               # save the figure data into the buffer
+    # check for the appropriate filename
+    if data["filename"] == "fig.png":
+        # plot a line with slope "a"
+        plt.plot([0, data["params"]["a"]], [0, 1])
+        # make a bytes object (a buffer)
+        buf = io.BytesIO()
+        # save the figure data into the buffer
+        plt.savefig(buf, format="png")
         return buf
 ```
 
 You can also use this functionality in file-based elements (`pl-figure`, `pl-file-download`) by setting `type="dynamic"`.
+
+## Custom grading best practices
+
+Although questions with custom grading usually don't use the grading functions from individual elements, it is _highly_ recommended that built-in elements are used for student input, as these elements include helpful parsing and feedback by default. Parsed student answers are present in the `data["submitted_answers"]` dictionary.
+
+!!! note
+
+    Data stored under the `"submitted_answers"` key in the data dictionary may be of varying type. Specifically, the `pl-integer-input`
+    element sometimes stores very large integers as strings instead of the Python `int` type used in most cases. The best practice for custom grader
+    code in this case is to always cast the data to the desired type, for example `int(data["submitted_answers"][name])`. See the
+    [PrairieLearn elements documentation](../elements.md) for more detailed discussion related to specific elements.
+
+### Recalculating scores
+
+Any custom grading function for the whole question should set `data["score"]` as a value between 0.0 and 1.0, which will be the final score for the given question. If a custom grading function is only grading a specific part of a question, the grading function should set the corresponding dictionary entry in `data["partial_scores"]` and then recompute the final `data["score"]` value for the whole question. The `prairielearn` Python library provides the following score recomputation functions:
+
+- [`set_weighted_score_data`][prairielearn.question_utils.set_weighted_score_data] sets `data["score"]` to be the weighted average of entries in `data["partial_scores"]`.
+- [`set_all_or_nothing_score_data`][prairielearn.question_utils.set_all_or_nothing_score_data] sets `data["score"]` to 1.0 if all entries in `data["partial_scores"]` are 1.0, 0.0 otherwise.
+
+This can be used like so:
+
+```python title="server.py"
+from prairielearn import set_weighted_score_data
+
+def grade(data):
+    # update partial_scores as necessary
+    # ...
+
+    # compute total question score
+    set_weighted_score_data(data)
```

Comment:
Nit: if this is a best practice, can we update the example above with this?

Reply:
8de961fa8bacdf280e06dd8f5e73d124836d5b15